source/fab/tools/tool_repository.py

```diff
@@ -0,0 +1,142 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+'''This file contains the ToolRepository class.
+'''
+
+# We can't declare _singleton and __new__() using ToolRepository, but
+# it is allowed if we use this import:
+from __future__ import annotations
+
+import logging
+from typing import Any, Type
+
+from fab.tools.tool import Tool
+from fab.tools.categories import Categories
+from fab.tools.linker import Linker
+from fab.tools.versioning import Fcm, Git, Subversion
+
+
+class ToolRepository(dict):
+    '''This class implements the tool repository. It stores a list of
+    tools for various categories. For each compiler, it will automatically
+    create a tool called "linker-{compiler-name}" which can be used for
+    linking with the specified compiler.
+    '''
+
+    _singleton: None | ToolRepository = None
+
+    def __new__(cls) -> ToolRepository:
+        '''Singleton access. Changes the value of _singleton so that the
+        constructor can verify that it is indeed called from here.
+        '''
+        if not cls._singleton:
+            cls._singleton = super().__new__(cls)
+
+        return cls._singleton
+
+    def __init__(self):
+        # Note that in this singleton pattern the constructor is called each
+        # time the instance is requested (since we overwrite __new__). But
+        # we only want to initialise the instance once, so let the constructor
+        # not do anything if the singleton already exists:
+        if ToolRepository._singleton:
+            return
```

Comment:
The only other way I am aware of is using (e,g.) `ToolRepository.get()` - which I had originally :)  And I can add a test in the constructor that the singleton instance is still None (or maybe set a custom value to guarantee the constructor is only called from the static singleton getter)?

Reply:
Singleton seems to be a pattern which depends on data hiding for slick function. Something which Python deliberately eschews. Lets leave it as it is for the moment and see how we feel about it later.

Reply:
I am aware that in general Singleton in Python is disliked/not recommended, so I am happy to discuss this later. FWIW, I like singleton because it gives a single function (the constructor)which can be used to setup an instance. I might be missing something, but without this we would have to have a test in any function to make sure the class/object/module/package (whatever) is correctly set up.